File: src/scene/SplashScene.java
package scene;

import org.newdawn.slick.Color;
import org.newdawn.slick.GameContainer;
//import org.newdawn.slick.Image;
import org.newdawn.slick.SlickException;
import org.newdawn.slick.state.StateBasedGame;

import component.AnimatingColor;
import component.Panel;
import component.Positionable.ReferencePoint;
import component.sprite.Sprite;

import core.GameDirector;
import core.ImageStore;

/**
 * A Scene with a short animated introduction displaying the null && void team
 * name.
 */
public class SplashScene extends Scene {
	public static final SceneID ID = SceneID.SPLASH;
	
	private static final int WAIT_TIME = 3000;
	private static final int DELAY = 100;
	private int time;
	
	private Panel whitePanel;
	private Panel blackPanel;
	private AnimatingColor whitePanelAnimatingColor;
	private AnimatingColor blackPanelAnimatingColor;
	private Sprite nullLogo;
	private Sprite voidLogo;

	@Override
	public void init(GameContainer container, StateBasedGame game) throws SlickException {
		super.init(container,  game);
		
		whitePanel = new Panel(container, container.getWidth() / 2, container.getHeight());
		whitePanelAnimatingColor = new AnimatingColor(Color.black, Color.white, WAIT_TIME / 4 - DELAY);
		whitePanel.setBackgroundColor(whitePanelAnimatingColor);
		backgroundLayer.add(whitePanel, mainLayer.getPosition(ReferencePoint.TOPLEFT), ReferencePoint.TOPLEFT);
		
		blackPanel = new Panel(container, container.getWidth() / 2, container.getHeight());
		blackPanelAnimatingColor = new AnimatingColor(Color.black, Color.white, WAIT_TIME / 4 - DELAY);
		blackPanel.setBackgroundColor(blackPanelAnimatingColor);
		backgroundLayer.add(blackPanel, mainLayer.getPosition(ReferencePoint.TOPRIGHT), ReferencePoint.TOPRIGHT);
		
		nullLogo = new Sprite(container, ImageStore.get().IMAGES.get("NULL")); //new Image(ConstantStore.PATH_LOGO + "null.png", false, Image.FILTER_NEAREST));
		whitePanel.add(nullLogo, whitePanel.getPosition(ReferencePoint.CENTERRIGHT), ReferencePoint.CENTERRIGHT, -10, 0);
		nullLogo.setVisible(false);
		
		voidLogo = new Sprite(container, ImageStore.get().IMAGES.get("VOID"));  //new Image(ConstantStore.PATH_LOGO + "void.png", false, Image.FILTER_NEAREST));
		blackPanel.add(voidLogo, whitePanel.getPosition(ReferencePoint.CENTERRIGHT), ReferencePoint.CENTERLEFT, 10, 0);
		voidLogo.setVisible(false);
		
		time = 0;
	}
	
	@Override
	public void update(GameContainer container, StateBasedGame game, int delta)
			throws SlickException {
		time += delta;
		
		if (time >= WAIT_TIME) {
			GameDirector.sharedSceneListener().requestScene(SceneID.LOADING, this, true);
		} else if (time >= WAIT_TIME * 1 / 2 + DELAY) {
			voidLogo.setVisible(true);
			
			blackPanelAnimatingColor.update(delta);
		} else if (time >= WAIT_TIME / 4 + DELAY) {
			nullLogo.setVisible(true);
			whitePanelAnimatingColor.update(delta);
			
			blackPanelAnimatingColor = new AnimatingColor(Color.white, Color.black, WAIT_TIME / 4 - DELAY);
			blackPanel.setBackgroundColor(blackPanelAnimatingColor);
		} else if (time < WAIT_TIME / 4) {
			blackPanelAnimatingColor.update(delta);
		} 
	}

	@Override
	public int getID() {
		return ID.ordinal();
	}
}

File: web/src/scene/SplashScene.ts
import { Scene } from './Scene';
import { SceneID } from './SceneID';
import { Panel } from '../component/Panel';
import { Sprite } from '../component/Sprite';
import { ReferencePoint } from '../component/Component';
import { AnimatingColor } from '../core/AnimatingColor';
import { Color } from '../core/Color';
import { ImageStore } from '../core/ImageStore';

export class SplashScene extends Scene {
  private static readonly WAIT_TIME = 3000;
  private static readonly DELAY = 100;

  private time: number = 0;
  private whitePanel!: Panel;
  private blackPanel!: Panel;
  private whitePanelAnimatingColor!: AnimatingColor;
  private blackPanelAnimatingColor!: AnimatingColor;
  private nullLogo!: Sprite;
  private voidLogo!: Sprite;
  private onComplete: (() => void) | null = null;

  constructor(canvasWidth: number, canvasHeight: number, onComplete: () => void) {
    super();
    this.onComplete = onComplete;
    this.init(canvasWidth, canvasHeight);
  }

  private init(canvasWidth: number, canvasHeight: number): void {
    this.whitePanel = new Panel(canvasWidth / 2, canvasHeight);
    this.whitePanelAnimatingColor = new AnimatingColor(
      Color.black,
      Color.white,
      SplashScene.WAIT_TIME / 4 - SplashScene.DELAY
    );
    this.whitePanel.setBackgroundColor(this.whitePanelAnimatingColor);
    this.backgroundLayer.add(this.whitePanel);
    this.whitePanel.setPosition({ x: 0, y: 0 }, ReferencePoint.TOPLEFT);

    this.blackPanel = new Panel(canvasWidth / 2, canvasHeight);
    this.blackPanelAnimatingColor = new AnimatingColor(
      Color.black,
      Color.white,
      SplashScene.WAIT_TIME / 4 - SplashScene.DELAY
    );
    this.blackPanel.setBackgroundColor(this.blackPanelAnimatingColor);
    this.backgroundLayer.add(this.blackPanel);
    this.blackPanel.setPosition({ x: canvasWidth / 2, y: 0 }, ReferencePoint.TOPLEFT);

    const nullImage = ImageStore.getImage('NULL');
    this.nullLogo = new Sprite(nullImage.width, nullImage.height, nullImage);
    this.whitePanel.add(
      this.nullLogo,
      this.whitePanel.getPosition(ReferencePoint.CENTERRIGHT),
      ReferencePoint.CENTERRIGHT,
      -10,
      0
    );
    this.nullLogo.setVisible(false);

    const voidImage = ImageStore.getImage('VOID');
    this.voidLogo = new Sprite(voidImage.width, voidImage.height, voidImage);
    this.blackPanel.add(
      this.voidLogo,
      this.blackPanel.getPosition(ReferencePoint.CENTERLEFT),
      ReferencePoint.CENTERLEFT,
      10,
      0
    );
    this.voidLogo.setVisible(false);
  }

  override update(delta: number): void {
    super.update(delta);

    this.time += delta;

    if (this.time >= SplashScene.WAIT_TIME) {
      if (this.onComplete) {
        this.onComplete();
        this.onComplete = null;
      }
    } else if (this.time >= (SplashScene.WAIT_TIME * 1) / 2 + SplashScene.DELAY) {
      this.voidLogo.setVisible(true);
      this.blackPanelAnimatingColor.update(delta);
    } else if (this.time >= SplashScene.WAIT_TIME / 4 + SplashScene.DELAY) {
      this.nullLogo.setVisible(true);
      this.whitePanelAnimatingColor.update(delta);

      this.blackPanelAnimatingColor = new AnimatingColor(
        Color.white,
        Color.black,
        SplashScene.WAIT_TIME / 4 - SplashScene.DELAY
      );
      this.blackPanel.setBackgroundColor(this.blackPanelAnimatingColor);
    } else if (this.time < SplashScene.WAIT_TIME / 4) {
      this.blackPanelAnimatingColor.update(delta);
    }
  }

  getID(): SceneID {
    return SceneID.SPLASH;
  }
}
